File: advent-of-code-2018/12.rb
require 'set'

INITIAL = '##.##..#.#....#.##...###.##.#.#..###.#....##.###.#..###...#.##.#...#.#####.###.##..#######.####..#'
ALIVE = File.foreach('12_input.txt').drop(2).map(&:split).group_by(&:last)['#'].map(&:first).to_set

def solve(gens)
  d = INITIAL.length * 10
  cushion = '.' * d
  initial = cushion + INITIAL + cushion

  gens.times.reduce({ state: initial, sums: [] }) { |acc, _|
    state = (2...initial.length - 2).reduce('..') { |s, i|
      s + (ALIVE.include?(acc[:state][i - 2..i + 2]) ? '#' : '.')
    } + '..'

    sum = state.chars.each_with_index.map { |c, i| c == '#' ? i - d : 0 }.sum

    { state: state, sums: [*acc[:sums], sum] }
  }
end

gens = 200
sums = solve(gens)[:sums]

p sums[19]

convergence = sums[1..-1].zip(sums).map { |x, y| x - y }.last

p sums[gens - 1] + convergence * (50_000_000_000 - gens)

File: advent-of-code-2018/12.hs
import Control.Monad.State
import Data.List (foldr, partition)
import Data.Set (Set (..), fromList, member)

alive :: IO (Set String)
alive = readFile "12_input.txt" >>= return . fromList . map (take 5) . fst . partition ((== '#') . last) . drop 2 . lines

initial :: String
initial = "##.##..#.#....#.##...###.##.#.#..###.#....##.###.#..###...#.##.#...#.#####.###.##..#######.####..#"

grow :: (Set String) -> String -> Int -> Char
grow alive s n =
  let spread = take 5 $ drop n s
   in if member spread alive then '#' else '.'

tick :: (Set String) -> Int -> State String Int
tick alive d = do
  s <- get
  let s' = ".." ++ foldr (\n acc -> (grow alive s n) : acc) ".." [0..length s]
  put s'
  return . sum $ zipWith (\c i -> if c == '#' then i else 0) s' [-d..]

generate :: Int -> IO [Int]
generate n = do
  a <- alive
  let d = length initial * 10
  let cushion = replicate d '.'
  let s = cushion ++ initial ++ cushion
  return $ evalState (sequence . replicate n $ tick a d) s

partOne :: IO Int
partOne = generate 20 >>= return . last

main :: IO ()
main = do
  partOne >>= print
